art/attacks/evasion/wasserstein.py

```diff
@@ -0,0 +1,1151 @@
+        :type eps_iter: `int`
+        :param eps_factor: Factor to increase the epsilon.
+        :type eps_factor: `float`
+        :param eps_step: Attack step size (input variation) at each iteration.
+        :type eps_step: `float`
+        :param regularization: Entropy regularization.
+        :type regularization: `float`
+        :param conjugate_sinkhorn_max_iter: The maximum number of iterations for the conjugate sinkhorn optimizer.
+        :type conjugate_sinkhorn_max_iter: `int`
+        :param projected_sinkhorn_max_iter: The maximum number of iterations for the projected sinkhorn optimizer.
+        :type projected_sinkhorn_max_iter: `int`
+        :param batch_size: Size of batches.
+        :type batch_size: `int`
+        :return: Adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        adv_x = x.copy().astype(ART_NUMPY_DTYPE)
+        adv_x_best = x.copy().astype(ART_NUMPY_DTYPE)
+
+        if targeted:
+            err = np.argmax(
+                self.estimator.predict(adv_x, batch_size=batch_size),
+                axis=1
+            ) == np.argmax(targets, axis=1)
+
+        else:
+            err = np.argmax(
+                self.estimator.predict(adv_x, batch_size=batch_size),
+                axis=1
+            ) != np.argmax(targets, axis=1)
+
+        err_best = err
+        eps = np.ones(batch_size) * eps
+
+        for i in range(max_iter):
+            adv_x = self._compute(
+                adv_x,
+                x,
+                targets,
+                cost_matrix,
+                kernel_size,
+                norm,
+                ball,
+                targeted,
+                eps,
+                eps_step,
+                regularization,
+                conjugate_sinkhorn_max_iter,
+                projected_sinkhorn_max_iter,
+                batch_size,
+                err,
+            )
+
+            if targeted:
+                err = np.argmax(
+                    self.estimator.predict(adv_x, batch_size=batch_size),
+                    axis=1
+                ) == np.argmax(targets, axis=1)
+
+            else:
+                err = np.argmax(
+                    self.estimator.predict(adv_x, batch_size=batch_size),
+                    axis=1
+                ) != np.argmax(targets, axis=1)
+
+            if np.mean(err) > np.mean(err_best):
+                err_best = err
+                adv_x_best = adv_x.copy()
+
+            if np.mean(err) == 1:
+                break
+
+            if (i + 1) % eps_iter == 0:
+                eps[~err] *= eps_factor
+
+        return adv_x_best
+
+    def _compute(
+            self,
+            x_adv,
+            x_init,
+            y,
+            cost_matrix,
+            kernel_size,
+            norm,
+            ball,
+            targeted,
+            eps,
+            eps_step,
+            regularization,
+            conjugate_sinkhorn_max_iter,
+            projected_sinkhorn_max_iter,
```

Comment:
Are these arguments required? Could they be replaced by direct access to self inside self._generate_batch?

Reply:
See comments above.